### ISSUE TITLE
Prefered way to do batch training of LSTM

### ISSUE THREAD

**Issue by lancerkz** [fchollet/keras#1024]
Hi All
I'm new to Keras. but this is indeed an amazing framework. Thanks to you all!
Due to hardware(RAM) limitation, I cannot load all training data to RAM. so I have to do batch training. 
After initial investigation of tickets here. seems I can do batch trainings in following two ways. but I'm not sure whether one of them is preferred. so any comment is welcome if you guys have similar situation. thanks ! 

1, using train_on_batch

        model = self.createLSTM()   #define my own model
        nb_epoch = 50
        batch_size = 128

        for e in range(nb_epoch):
            self.logger.info("Epoch:%d" % (e+1))
            for idx, f in enumerate(vlist):               
                samples, labels = self.loadData(f)  #large training data 3-4G each 
                nb_batch = len(samples)/batch_size
                for idy in xrange(nb_batch):
                    start = idy * batch_size
                    if idy == nb_batch - 1:
                        end = len(samples)
                    else:
                        end = start + batch_size

                    batch_samples = samples[start:end,:]
                    batch_labels = labels[start:end] 
                    loss, acc = model.train_on_batch(batch_samples, batch_labels, accuracy=True)
   
2, using fit directly, and set nb_epoch as 1

        model = self.createLSTM()   #define my own model
        nb_epoch = 50
        batch_size = 128

        for e in range(nb_epoch):
            self.logger.info("Epoch:%d" % (e+1))
            for idx, f in enumerate(vlist):     
                samples, labels = self.loadData(f)  #large training data 3-4G each 
                model.fit(samples, labels, nb_epoch=1, validation_split=0.2, show_accuracy=True)

I'm not sure which one is better. To me, "fit" interface provides validation and batch functionality, so it's better. any comment? 

Kenneth

**Comment by EderSantana** [comment]
if you can't fit your data in memory, maybe your should try HDF5Matrix class to load your dataset.


Note for the future: this is the third time I recently answer a question which the solutions is basically how to use HDF5Matrix. Write a simple example that will catch people's eyes.

**Comment by lancerkz** [comment]
@EderSantana 
Thanks for quick response! If I understand correctly, HDF5Matrix is used to dump/load data which is indeed problem in batch training. but it's not the major problem I concerned. 
Actually, I'm using numpy.savez/load to dump/load training data. speed seems acceptable to me. 
Back to the topic, Simply speaking, my question is "Is there any reason that I should't use fit function like that in batch training in a real job?" 

Kenneth

**Comment by EderSantana** [comment]
fit simply call `_train_on_batch` batch by batch. you can do that manually on you own for loop. Using fit just makes your code look cleaner, but it is not mandatory at all

**Comment by fchollet** [comment]
Both the methods you mention are largely equivalent, I wouldn't say one is better than another. Maybe `train_on_batch` is somewhat easier to understand, though. Take your pick...

**Comment by arisliang** [comment]
Does LSTM models need to be stateful in this case? Will the weights reset when train across batches?

**Comment by NarenSuri** [comment]
I have the same question, can someone answer this

> Does LSTM models need to be stateful in this case? Will the weights reset when train across batches?
